### PR TITLE
Extract BackendPool interface into three 3 separate interfaces

### PR DESCRIPTION
**Pull request by rramkumar1**
This PR extracts the BackendPool interface into 3 separate interfaces:

    1. BackendPool - Simple CRUD for backends
    2. Syncer - Sync and GC backends
    3. Linker - Link backends to their associated groups. 

This allows us to break the current BackendPool implementation up into more manageable chunks and each chunk is much more testable. Furthermore, some implementations, namely BackendPool and Syncer, can be easily reused. 

/assign @bowei

### CODE REVIEW

File: pkg/backends/ig_linker.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
+package backends
```
Comment (bowei) [inline pkg/backends/ig_linker.go]: copyright

Comment (rramkumar1) [reply pkg/backends/ig_linker.go]: Fixed everywhere.

**Comment by rramkumar1** [comment]
All tests are now migrated over to new interfaces. 

Added a new test file called integration_test.go which contains some tests carried over from the original. Those are failing so need to figure out why.

**Comment by rramkumar1** [comment]
Update: All tests pass.

File: pkg/backends/backends.go [hunk 2]
```diff
@@ -158,243 +78,63 @@ func ensureDescription(be *composite.BackendService, sp *utils.ServicePort) (nee
...
-		return nil
 	}
-	return fmt.Errorf("received errors when updating backend service: %v", strings.Join(errs, "\n"))
+	b.snapshotter.Add(name, be)
+	return be, nil
 }
 
-// Delete deletes the Backend for the given port.
+// Implements Pool.
 func (b *Backends) Delete(name string) (err error) {
```
Comment (bowei) [inline pkg/backends/backends.go]: Delete implements ...

Comment (bowei) [reply pkg/backends/backends.go]: Can we use non-implicit `err` here?

File: pkg/backends/backends.go [hunk 3]
```diff
@@ -411,208 +151,47 @@ func (b *Backends) Delete(name string) (err error) {
...
-		oldBackends.Insert(be.Group)
-	}
-	for _, be := range targetBackends {
-		newBackends.Insert(be.Group)
-	}
-
-	if !oldBackends.Equal(newBackends) {
-		backendService.Backends = targetBackends
-		return b.cloud.UpdateBetaGlobalBackendService(backendService)
+// Implements Pool.
```
Comment (bowei) [inline pkg/backends/backends.go]: GetLocalSnapshot implements Pool.

File: pkg/backends/backends.go [hunk 4]
```diff
@@ -158,243 +78,63 @@ func ensureDescription(be *composite.BackendService, sp *utils.ServicePort) (nee
...
+	b.snapshotter.Add(be.Name, be)
 	return nil
 }
 
-// edgeHop checks the links of the given backend by executing an edge hop.
-// It fixes broken links and updates the Backend accordingly.
-func (b *Backends) edgeHop(be *composite.BackendService, igLinks []string) error {
-	addIGs, err := getInstanceGroupsToAdd(be, igLinks)
+// Implements Pool.
+func (b *Backends) Get(name string, version meta.Version) (*composite.BackendService, error) {
```
Comment (bowei) [inline pkg/backends/backends.go]: Get implements ...

File: pkg/backends/backends.go [hunk 5]
```diff
@@ -411,208 +151,47 @@ func (b *Backends) Delete(name string) (err error) {
...
-	}
-	return nil
+	return
 }
 
-// Status returns the status of the given backend by name.
-func (b *Backends) Status(name string) string {
-	backend, err := b.cloud.GetGlobalBackendService(name)
-	if err != nil || len(backend.Backends) == 0 {
+// Implements Pool.
```
Comment (bowei) [inline pkg/backends/backends.go]: // Health implements ..

File: pkg/backends/ig_linker.go [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+// maxRPS is the RPS setting for all Backends with BalancingMode RATE. The exact
+// value doesn't matter, as long as it's the same for all Backends. Requests
+// received by GCLB above this RPS are NOT dropped, GCLB continues to distribute
+// them across IGs.
+// TODO: Should this be math.MaxInt64?
+const maxRPS = 1
+
+// InstanceGroupLinker handles linking backends to InstanceGroup's.
+type InstanceGroupLinker struct {
```
Comment (bowei) [inline pkg/backends/ig_linker.go]: does this have to be exported?

File: pkg/backends/ig_linker.go [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+	backendPool Pool,
+	namer *utils.Namer) *InstanceGroupLinker {
+	return &InstanceGroupLinker{
+		instancePool: instancePool,
+		backendPool:  backendPool,
+		namer:        namer,
+	}
+}
+
+// Implements Link.
```
Comment (bowei) [inline pkg/backends/ig_linker.go]: fix all comments

File: pkg/backends/ig_linker.go [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+	originalIGBackends := []*composite.Backend{}
+	for _, backend := range be.Backends {
+		// Backend service is not able to point to NEG and IG at the same time.
+		// Filter IG backends here.
+		if strings.Contains(backend.Group, "instanceGroups") {
+			originalIGBackends = append(originalIGBackends, backend)
+		}
+	}
+
+	// We first try to create the backend with balancingMode=RATE.  If this	+	return addIGs
```
Comment (bowei) [inline pkg/backends/ig_linker.go]: weird formatting on this line

File: pkg/backends/integration_test.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
Comment (bowei) [inline pkg/backends/integration_test.go]: date

File: pkg/backends/ig_linker.go (hunk 7 above)
Comment (rramkumar1) [reply pkg/backends/ig_linker.go]: Done

File: pkg/backends/ig_linker.go (hunk 8 above)
Comment (rramkumar1) [reply pkg/backends/ig_linker.go]: Done

File: pkg/backends/integration_test.go (hunk 9 above)
Comment (rramkumar1) [reply pkg/backends/integration_test.go]: Done

File: pkg/backends/backends.go (hunk 5 above)
Comment (rramkumar1) [reply pkg/backends/backends.go]: Done

File: pkg/backends/backends.go (hunk 4 above)
Comment (rramkumar1) [reply pkg/backends/backends.go]: Done

File: pkg/backends/backends.go (hunk 3 above)
Comment (rramkumar1) [reply pkg/backends/backends.go]: Done

File: pkg/backends/backends.go [hunk 10]
```diff
@@ -158,243 +78,63 @@ func ensureDescription(be *composite.BackendService, sp *utils.ServicePort) (nee
...
-		if strings.Contains(existingHCLink, "/httpHealthChecks/") {
-			hasLegacyHC = true
-		}
-	}
-
-	// Ensure health check for backend service exists. Note that hasLegacyHC
-	// will dictate whether we search for an existing legacy health check.
-	hcLink, err := b.ensureHealthCheck(sp, hasLegacyHC)
-	if err != nil {
+// Implements Pool.
```
[the remark this answers is not in the document]
Comment (rramkumar1) [reply pkg/backends/backends.go]: Done

File: pkg/backends/backends.go (hunk 2 above)
Comment (rramkumar1) [reply pkg/backends/backends.go]: I wasn't able to get rid of it because of the defer.

File: pkg/backends/ig_linker.go (hunk 6 above)
Comment (rramkumar1) [reply pkg/backends/ig_linker.go]: Fixed it for this and others as well.

**Pull request merged**